CircuitPython_Pyloton/pyloton.py

```diff
@@ -0,0 +1,563 @@
+from adafruit_bitmap_font import bitmap_font
+from adafruit_display_shapes.rect import Rect
+from adafruit_display_text import label
+from adafruit_ble_apple_media import AppleMediaService
+from adafruit_ble_apple_media import UnsupportedCommand
+import gamepad
+import touchio
+
+class Clue:
+    """
+    A very minimal version of the CLUE library.
+    The library requires the use of many sensor-specific
+    libraries this project doesn't use, and they were
+    taking up a lot of RAM.
+    """
+    def __init__(self):
+        self._i2c = board.I2C()
+
+        self._touches = [board.D0, board.D1, board.D2]
+        self._touch_threshold_adjustment = 0
+
+        self._a = digitalio.DigitalInOut(board.BUTTON_A)
+        self._a.switch_to_input(pull=digitalio.Pull.UP)
+        self._b = digitalio.DigitalInOut(board.BUTTON_B)
+        self._b.switch_to_input(pull=digitalio.Pull.UP)
+        self._gamepad = gamepad.GamePad(self._a, self._b)
+
+    @property
+    def were_pressed(self):
+        """
+        Returns a set of buttons that have been pressed since the last time were_pressed was run.
+        """
+        ret = set()
+        pressed = self._gamepad.get_pressed()
+        for button, mask in (('A', 0x01), ('B', 0x02)):
+            if mask & pressed:
+                ret.add(button)
+        return ret
+
+    def _touch(self, i):
+        if not isinstance(self._touches[i], touchio.TouchIn):
+            self._touches[i] = touchio.TouchIn(self._touches[i])
+            self._touches[i].threshold += self._touch_threshold_adjustment
+        return self._touches[i].value
+
+    @property
+    def touch_0(self):
+        """
+        Returns True when capacitive touchpad 0 is currently being pressed.
+        """
+        return self._touch(0)
+
+    @property
+    def touch_1(self):
+        """
+        Returns True when capacitive touchpad 1 is currently being pressed.
+        """
+        return self._touch(1)
+
+    @property
+    def touch_2(self):
+        """
+        Returns True when capacitive touchpad 2 is currently being pressed.
+        """
+        return self._touch(2)
+
+
+
+class Pyloton:
+    """
+    Contains the various functions necessary for doing the Pyloton learn guide.
+    """
+    # pylint: disable=too-many-instance-attributes
+    _previous_wheel = 0
+    _previous_crank = 0
+    _previous_revolutions = 0
+    _previous_rev = 0
+    _previous_speed = 0
+    _previous_cadence = 0
+    _previous_heart = 0
+    _i = 0
+    _j = 0
+    _setup = 0
+    _hr_label = None
+    _sp_label = None
+    _cad_label = None
+    _ams_label = None
+    _hr_service = None
+    _heart_y = None
+    _speed_y = None
+    _cad_y = None
+    _ams_y = None
+
+    ams = None
+    cyc_connections = None
+    cyc_services = None
+
+    splash = displayio.Group(max_size=25)
+    loading_group = displayio.Group()
+
+    YELLOW = 0xFCFF00
+    PURPLE = 0x64337E
+    WHITE = 0xFFFFFF
+
+    start = time.time()
+
+    track_artist = True
```

Comment:
Many of these are class variables, but they are used like instance variables. I'd try to be consistent and assign them in `__init__()`, or else don't make this a class (but a class is clearer).